netket/experimental/dynamics/_integrator.py

```diff
@@ -0,0 +1,436 @@
+        """
+        self.dt = dt
+
+        self.atol = atol
+        self.rtol = rtol
+        if dt_limits is None:
+            dt_limits = (None, 10 * dt)
+        self.dt_limits = dt_limits
+
+
+class Integrator(struct.Pytree, mutable=True):
+    r"""
+    Ordinary-Differential-Equation integrator.
+    Given an ODE-function :math:`dy/dt = f(t, y)`, it integrates the derivatives to obtain the solution
+    at the next time step :math:`y_{t+1}`.
+    """
+
+    f: Callable = struct.field(pytree_node=False)
+    """The ODE function."""
+
+    _state: IntegratorState
+    """The state of the integrator, containing informations about the solution."""
+    _solver: "AbstractSolver"
+    """The ODE solver."""
+
+    use_adaptive: bool = struct.field(pytree_node=False)
+    """Boolean indicating whether to use an adaptative scheme."""
+    norm: Callable = struct.field(pytree_node=False)
+    """The norm used to estimate the error."""
+
+    atol: float = struct.field(pytree_node=False)
+    """Absolute tolerance on the error of the state."""
+    rtol: float = struct.field(pytree_node=False)
+    """Relative tolerance on the error of the state."""
+    dt_limits: Optional[LimitsDType] = struct.field(pytree_node=False)
+    """Limits of the time-step size."""
+
+    _do_step: Callable = struct.field(pytree_node=False)
+    """The function that performs the time step."""
+
+    def __init__(
+        self,
+        f: Callable,
+        solver: "AbstractSolver",
+        t0: float,
+        y0: PyTree,
+        use_adaptive: bool,
+        parameters: IntegratorParameters,
+        norm: str | Callable = None,
+    ):
+        r"""
+        Args:
+            f: The ODE function
+                Given a time `t` and a state `y_t`, it should return the partial
+                derivatives in the same format as `y_t`. The function should also accept
+                supplementary arguments, such as :code:`stage`.
+            solver: The ODE solver.
+                If :code:`use_adaptive`, it should define the method :code:`step_with_error(f,dt,t,y_t,solver_state)`
+                otherwise, the method :code:`step(f,dt,t,y_t,solver_state)`
+            t0: The intial time.
+            y0: The initial state.
+            use_adaptive: The (boolean) indicator whether an adaptive (time-step) scheme is used.
+            parameters: The supplkementary hyper-parameters of the integrator.
+                This includes the values for :code:`dt`, :code:`atol`, :code`rtol` and :code`dt_limits`
+                See :code:`IntegratorParameters` for more details.
+            norm: The function used for the norm of the error.
+                By default, we use euclidean_norm.
+        """
+        self.f = f
+        self._solver = solver
+
+        self.use_adaptive = use_adaptive
+        if use_adaptive:
+            self._do_step = general_time_step_adaptive
+        else:
+            self._do_step = general_time_step_fixed
```

Comment:
```suggestion
```

let's move this switch to the `_step` function and simply call the right method. It would be more pythonic.